### PR TITLE
Reward system update

### PR DESCRIPTION
**Pull request by Cattharine**
- Updated reward system. Now it does not depend on state vector
- Added new reward system

### CODE REVIEW

File: environment/reward_sytem/empirical_reward_system.py [hunk 1]
```diff
@@ -11,9 +12,10 @@ def __init__(self, config: dict) -> None:
         self.lose_reward = -50
         self.remove_sprint_card_reward = -2
         self.valid_action_reward = 1
+        self.money_weight = 1e-3
 
-    def get_reward(self, state_old, action, state_new) -> float:
-        if (state_old == state_new).all():
```
Comment (krutovsky-danya) [inline environment/reward_sytem/empirical_reward_system.py:16]: Со словарями равенство в питоне хорошо работает, поэтому можно оставить этот if и не передавать значение success (мне кажется это костыль)

File: environment/reward_sytem/base_reward_system.py [hunk 2]
```diff
@@ -1,15 +1,33 @@
...
+        self.customers_index = "customers"
+        self.loyalty_index = "loyalty"
+        self.credit_index = "credit"
+        self.sprint_hours_index = "sprint_hours"
+        self.done_index = "done"
+        self.blank_sprint_counter_index = "blank_sprint_counter"
         self.config = config
 
-    def get_reward(self, state_old, action, state_new) -> float:
+    def copy_state(self, game: ProductOwnerGame):
```
Comment (krutovsky-danya) [inline environment/reward_sytem/base_reward_system.py]: Хорошая идея, но реализация не должна лежать в reward_system, можно положить ее отдельно. Например рядом с environment

File: environment/environment.py [hunk 3]
```diff
@@ -171,13 +175,14 @@ def _set_info_cards(self, cards, offset: int, predicate, result):
...
+        state_old = self.reward_system.copy_state(self.game)
         success = self._perform_action(action)
         if success:
             self.current_state = self._get_state()
-        reward = self.reward_system.get_reward(state_old, action, self.current_state)
+        state_new = self.reward_system.copy_state(self.game)
+        reward = self.reward_system.get_reward(state_old, action, state_new, success)
         info = self.get_info()
-        done = self.game.context.done or len(info) == 0
+        done = self.game.context.done or len(info) == 0 or (self.game.context.customers <= 0 and not self.game.context.is_new_game)
```
Comment (krutovsky-danya) [inline environment/environment.py]: Очень много логики в одной строчке, давай вынесем в отдельный метод, который можно разбить на несколько строк)

File: environment/reward_sytem/full_potential_credit_reward_system.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        self.money_weight = 1e-3
+
+    def get_reward(self, state_old, action, state_new, success) -> float:
+        reward = super().get_reward(state_old, action, state_new, success)
+        reward += self.get_credit_payer_reward(state_old, state_new)
+
+        return reward
+
+    def get_credit_payer_reward(self, state_old, state_new) -> float:
+        first = self.get_first_sum_part(state_new) - self.get_first_sum_part(state_old)
```
Comment (krutovsky-danya) [inline environment/reward_sytem/full_potential_credit_reward_system.py]: Названия first, second и так далее ничего не говорят о смысле переменной, нужно переименовать
Почему-то же ты разделила методы

File: environment/reward_sytem/full_potential_credit_reward_system.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+        reward = first + second + third + fourth
+
+        return reward
+
+    def get_first_sum_part(self, state) -> float:
+        blank_sprint_counter = self.get_blank_sprint_counter(state)
+        customers = self.get_customers(state)
+        loyalty = self.get_loyalty(state)
+        return max((7 - blank_sprint_counter) * customers * loyalty, 0) * self.potential_weight
```
Comment (krutovsky-danya) [inline environment/reward_sytem/full_potential_credit_reward_system.py]: `customers * loyalty * self.potential_weight` - это, видимо, доход за пустой спринт, можно назвать `income` 
А остальной множитель - количество таких спринтов, поэтому назвал бы такой метод get_free_income

File: environment/reward_sytem/full_potential_credit_reward_system.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    def get_credit_payer_reward(self, state_old, state_new) -> float:
+        first = self.get_first_sum_part(state_new) - self.get_first_sum_part(state_old)
+        # if self.get_sprint_hours(state_old) > 0 and first < 0:
+        #     first = 0
+        second = self.get_second_sum_part(state_new) - self.get_second_sum_part(state_old)
+        third = self.get_third_sum_part(state_new) - self.get_third_sum_part(state_old)
+        fourth = self.get_fourth_sum_part(state_new) - self.get_fourth_sum_part(state_old)
+
+        reward = first + second + third + fourth
```
Comment (krutovsky-danya) [inline environment/reward_sytem/full_potential_credit_reward_system.py]: В телеге было написано, что считается разница потенциальных доходов, можно сумму слагаемых и дописать как potential_income и брать разницу значений нового состояния и старого

File: environment/reward_sytem/full_potential_credit_reward_system.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        customers = self.get_customers(state)
+        loyalty = self.get_loyalty(state)
+        return max((7 - blank_sprint_counter) * customers * loyalty, 0) * self.potential_weight
+
+    def get_second_sum_part(self, state) -> float:
+        blank_counter = self.get_blank_sprint_counter(state)
+        customers = self.get_customers(state)
+        loyalty = self.get_loyalty(state)
+
+        if blank_counter - 4 < 9:
```
Comment (krutovsky-danya) [inline environment/reward_sytem/full_potential_credit_reward_system.py]: Как будто нужно инвертировать if + после return else не нужен

File: environment/reward_sytem/full_potential_credit_reward_system.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+            indexes = torch.arange(max(blank_counter - 4, 3), 9)
+            sums = (3 + indexes) / 2 * (indexes - 2) / 3
```
Comment (krutovsky-danya) [inline environment/reward_sytem/full_potential_credit_reward_system.py]: Непонятно, что такое sums по смыслу кажется, что это убывание величин, а не суммы

File: environment/reward_sytem/full_potential_credit_reward_system.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+            result = (customers - 0.5 * sums) * (loyalty - 0.05 * sums)
+            return torch.sum(result).item() * self.potential_weight
+        else:
+            return 0
+
+    def get_third_sum_part(self, state) -> float:
+        customers = self.get_customers(state)
+        loyalty = self.get_loyalty(state)
+
+        n = (2 * customers - 11) // 3
```
Comment (krutovsky-danya) [inline environment/reward_sytem/full_potential_credit_reward_system.py:76]: Магические константы хотелось бы назвать, если есть возможность. Если они например что-то значат, например, 11 точно откуда-то взялось, не из кода игры

**Review by krutovsky-danya — CHANGES_REQUESTED** [review]
Много докопался по стилю, по логике замечаний почти нет)

File: environment/reward_sytem/empirical_reward_system.py (hunk 1 above)
Comment (Cattharine) [reply environment/reward_sytem/empirical_reward_system.py:16]: Можно ли гарантировать, что равенство со словарями всегда будет соответствовать успеху действия? Даже если мы поменяем наполнение словаря когда-нибудь потом?

File: environment/reward_sytem/base_reward_system.py (hunk 2 above)
Comment (Cattharine) [reply environment/reward_sytem/base_reward_system.py]: Мне кажется, что удобно её расположить именно рядом с reward_system, чтобы знать, какая информация reward_system требуется для работы и как именно её получили

File: environment/reward_sytem/full_potential_credit_reward_system.py (hunk 7 above)
Comment (Cattharine) [reply environment/reward_sytem/full_potential_credit_reward_system.py]: Как инвертировать? blank_counter - 4 >=9 и действие под if то же? Нет, так как в этой части считается сумма от 7 до 13 пустого спринта

File: environment/reward_sytem/full_potential_credit_reward_system.py (hunk 8 above)
Comment (Cattharine) [reply environment/reward_sytem/full_potential_credit_reward_system.py]: Да, здесь нужен комментарий. Без деления на 3 это сумма арифметической прогрессии для каждого элемента indexes от 3 до элемента indexes включительно

File: environment/reward_sytem/full_potential_credit_reward_system.py (hunk 9 above)
Comment (Cattharine) [reply environment/reward_sytem/full_potential_credit_reward_system.py:76]: Тут скорее нужен комментарий, что переменная n является минимальным целым решением неравенства, числом шагов, за которое количество пользователей упадет ниже 0

File: environment/reward_sytem/empirical_reward_system.py (hunk 1 above)
Comment (krutovsky-danya) [reply environment/reward_sytem/empirical_reward_system.py:16]: Точнее говоря, равенство соответствует неуспеху
Потом в плане кода? Ну у нас равные состояния должны соответствовать одинаковым играм

File: environment/reward_sytem/base_reward_system.py (hunk 2 above)
Comment (krutovsky-danya) [reply environment/reward_sytem/base_reward_system.py]: Может по хорошему вообще игру научиться копировать?
Ну если рядом с reward_system, то точно не в самом классе, пожалуйста

File: environment/reward_sytem/full_potential_credit_reward_system.py (hunk 7 above)
Comment (krutovsky-danya) [reply environment/reward_sytem/full_potential_credit_reward_system.py]: Ну инвертировать условие ифа и тело else переставить 
```
if blank_counter - 4 >= 9:
  return 0
indexes = ...```

**Review by krutovsky-danya — CHANGES_REQUESTED** [review]

File: environment/reward_sytem/empirical_reward_system.py (hunk 1 above)
Comment (Cattharine) [reply environment/reward_sytem/empirical_reward_system.py:16]: Да, точно)
Но ведь словарь не полностью соответствует вектору состояния... 
К тому же я не считаю передачу success костылем. У нас есть прямой признак того, было действие совершено или нет. Почему бы его не использовать в системе наград вместо того, чтобы оценивать это по косвенным признакам?

File: environment/reward_sytem/base_reward_system.py (hunk 2 above)
Comment (Cattharine) [reply environment/reward_sytem/base_reward_system.py]: Думала над этим. Не хотелось бы копировать игру...
Тогда предлагаю так: создать ещё один класс, назвать его reward_adapter, перенести в него всю логику со словарем, унаследовать base_reward_system от reward_adapter. В таком случае системы наград задаются, начиная с base_reward_system, и всем системам наград будет доступна информация из адаптера

**Review by krutovsky-danya — APPROVED** [review]
